File: Components/WinForms/CS/EditorsMainDemo/TutorialControl.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.DXperience.Demos;
using DevExpress.Utils;
using DevExpress.XtraBars;
using DevExpress.XtraGrid;
using DevExpress.XtraTab;
using DevExpress.XtraVerticalGrid;

namespace DevExpress.XtraEditors.Demos {
    public class TutorialControl : TutorialControlBase {
        List<Panel> centerControls = new List<Panel>();
        protected virtual bool AllowCenterControls { get { return true; } }
        public TutorialControl() {
            this.Padding = new Padding(5, 0, 0, 5);
            this.SizeChanged += (s, e) => UpdatePosition();
            SubscribeOnEvents();
        }
        void UpdatePosition() {
            foreach(Panel pnl in centerControls) {
                if(pnl.Parent == null) 
                    continue;
                pnl.Location = new Point((pnl.Parent.Width - pnl.Width) / 2, (pnl.Parent.Height - pnl.Height) / 2);
            }
        }
        protected override void SetControlManager(Control ctrl, BarManager manager) {
            GridControl grid = ctrl as GridControl;
            if(grid != null) grid.MenuManager = manager;
            PropertyGridControl pGrid = ctrl as PropertyGridControl;
            if(pGrid != null) pGrid.MenuManager = manager;
            BaseEdit be = ctrl as BaseEdit;
            if(be != null) be.MenuManager = manager;
        }
        public override bool AllowPrintOptions { get { return false; } }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            if(AllowCenterControls) {
                foreach(Control ctrl in this.Controls) {
                    XtraTabControl tabControl = ctrl as XtraTabControl;
                    if(tabControl != null) UpdateTabControlPosition(tabControl);
                }
                UpdatePosition();
            }
            CalcContentCore();
        }
        protected virtual void CalcContentCore() { }
        protected virtual void OnLookAndFeelChanged(object sender, EventArgs e) {
            CalcContentCore();
        }
        protected virtual void SubscribeOnEvents() {
            if(LookAndFeel != null)
                LookAndFeel.StyleChanged += OnLookAndFeelChanged;
        }
        protected virtual void UnsubscribeOnEvents() {
            if(LookAndFeel != null)
                LookAndFeel.StyleChanged -= OnLookAndFeelChanged;
        }
        protected override void OnSizeChanged(EventArgs e) {
            base.OnSizeChanged(e);
            CalcContentCore();
        }
        protected override void OnVisibleChanged(EventArgs e) {
            base.OnVisibleChanged(e);
            if(Visible) CalcContentCore();
        }
        protected void CalcLayoutControlBounds(DevExpress.XtraLayout.LayoutControl layout) {
            if(layout == null || layout.Parent == null || !layout.Parent.IsHandleCreated)
                return;
            Size bestSize = layout.GetPreferredSize(Size.Empty);
            int bestWidth = Math.Max(0, Math.Min(layout.Parent.Width, bestSize.Width));
            int bestHeight = Math.Max(0, Math.Min(layout.Parent.Height, bestSize.Height));
            layout.SetBounds((layout.Parent.Width - bestWidth) / 2, (layout.Parent.Height - bestHeight) / 2, bestWidth, bestHeight);
        }

        protected virtual void UpdateTabControlPosition(XtraTabControl tabControl) {
            tabControl.SelectedPageChanged += (s, e) => UpdatePosition();
            foreach(XtraTabPage page in tabControl.TabPages) {
                if(page != null) UpdatePagePosition(page);
            }
        }

        void UpdatePagePosition(XtraTabPage page) {
            Panel pnl = new Panel();
            int left = 100, right = 0, top = 100, bottom = 0;
            for(int i = page.Controls.Count - 1; i >= 0; i--) {
                Control ctrl = page.Controls[i];
                if(ctrl is PopupContainerControl || ctrl is PopupControlContainer || ctrl is FlyoutPanel) continue;
                if(left > ctrl.Left) left = ctrl.Left;
                if(right < ctrl.Right) right = ctrl.Right;
                if(top > ctrl.Top) top = ctrl.Top;
                if(bottom < ctrl.Bottom) bottom = ctrl.Bottom;
                ctrl.Parent = pnl;
            }
            pnl.Parent = page;
            pnl.Size = new Size(right + left, bottom + top);
            centerControls.Add(pnl);
        }
        protected override void Dispose(bool disposing) {
            UnsubscribeOnEvents();
            base.Dispose(disposing);
        }
    }
}

File: Components/WinForms/VB/EditorsMainDemo/TutorialControl.vb
Imports System.Drawing.Imaging
Imports DevExpress.DXperience.Demos
Imports DevExpress.Utils.Menu
Imports DevExpress.XtraBars
Imports DevExpress.XtraTab
Imports DevExpress.XtraGrid
Imports DevExpress.XtraVerticalGrid
Imports DevExpress.Utils

Namespace DevExpress.XtraEditors.Demos
    Public Class TutorialControl
        Inherits TutorialControlBase

        Private centerControls As New List(Of Panel)()
        Protected Overridable ReadOnly Property AllowCenterControls() As Boolean
            Get
                Return True
            End Get
        End Property
        Public Sub New()
            Me.Padding = New Padding(5, 0, 0, 5)
            AddHandler Me.SizeChanged, AddressOf ControlSizeChanged
        End Sub
        Sub ControlSizeChanged(sender As Object, e As EventArgs)
            UpdatePosition()
        End Sub
        Private Sub UpdatePosition()
            For Each pnl As Panel In centerControls
                If (pnl.Parent Is Nothing) Then
                    Continue For
                End If
                pnl.Location = New Point((pnl.Parent.Width - pnl.Width) \ 2, (pnl.Parent.Height - pnl.Height) \ 2)
            Next pnl
        End Sub
        Protected Overrides Sub SetControlManager(ByVal ctrl As Control, ByVal manager As BarManager)
            Dim grid As GridControl = TryCast(ctrl, GridControl)
            If grid IsNot Nothing Then
                grid.MenuManager = manager
            End If
            Dim pGrid As PropertyGridControl = TryCast(ctrl, PropertyGridControl)
            If pGrid IsNot Nothing Then
                pGrid.MenuManager = manager
            End If
            Dim be As BaseEdit = TryCast(ctrl, BaseEdit)
            If be IsNot Nothing Then
                be.MenuManager = manager
            End If
        End Sub
        Public Overrides ReadOnly Property AllowPrintOptions() As Boolean
            Get
                Return False
            End Get
        End Property
        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            If AllowCenterControls Then
                For Each ctrl As Control In Me.Controls
                    Dim tabControl As XtraTabControl = TryCast(ctrl, XtraTabControl)
                    If tabControl IsNot Nothing Then
                        UpdateTabControlPosition(tabControl)
                    End If
                Next ctrl
                UpdatePosition()
            End If
        End Sub

        Private Sub UpdateTabControlPosition(ByVal tabControl As XtraTabControl)
            AddHandler tabControl.SelectedPageChanged, AddressOf ControlSizeChanged
            For Each page As XtraTabPage In tabControl.TabPages
                If page IsNot Nothing Then
                    UpdatePagePosition(page)
                End If
            Next page
        End Sub
        Private Sub UpdatePagePosition(ByVal page As XtraTabPage)
            Dim pnl As New Panel()
            Dim left As Integer = 100, right As Integer = 0, top As Integer = 100, bottom As Integer = 0
            For i As Integer = page.Controls.Count - 1 To 0 Step -1
                Dim ctrl As Control = page.Controls(i)
                If TypeOf ctrl Is PopupContainerControl OrElse TypeOf ctrl Is PopupControlContainer OrElse TypeOf ctrl Is FlyoutPanel Then
                    Continue For
                End If
                If left > ctrl.Left Then
                    left = ctrl.Left
                End If
                If right < ctrl.Right Then
                    right = ctrl.Right
                End If
                If top > ctrl.Top Then
                    top = ctrl.Top
                End If
                If bottom < ctrl.Bottom Then
                    bottom = ctrl.Bottom
                End If
                ctrl.Parent = pnl
            Next i
            pnl.Parent = page
            pnl.Size = New Size(right + left, bottom + top)
            centerControls.Add(pnl)
        End Sub
    End Class
End Namespace
